### ISSUE TITLE
[question] Why are `build_requires` version ranges of dependencies resolved?

### ISSUE THREAD

**Issue by mpll-d** [conan-io/conan#17463]
### What is your question?

Hello Conan people,
I'm currently experiencing (conan v2.10.2) quite slow `conan install` runs on my application (~8 min to run), most of the time being spent on the resolution of version ranges rather than binaries downloads.
I noticed that in particular, some of the libs I consume declare `cmake` and `ninja` as `tool_requires`, but my package does _not_.

I would suspect that, when building my app that does not list them as `tool_requires`, these packages would just be completely skipped, without spending time on resolving the revision across the various remotes.

I cannot fathom why Conan needs to actually look for them, could you clarify this? Otherwise, it would be very nice to optimize away this `tool_require` version range resolution when `conan install` is not called with `--build=missing` or similar options.

Thanks for your help,
Milan

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @mpll-d 

Thanks for your question.

Conan always expand the full dependency graph. That means that ``tool_requires`` recipes are evaluated always, the recipes might be downloaded, and then the existence of binaries will be evaluated. Most of the times, the **binaries** for those ``tool_requires`` are not necessary and they will be marked as "Skip", and the download will be avoided completely. But still the graph will be fully computed, this is the default because it avoids many issues, like lockfiles not being complete, later problems to reproduce a build because some missing tool-requires, etc. There are also situations in which users want their tool-requires to affect the produced binaries of the packages using those tool-requires.

For very large dependency graph (many thousands of nodes in the graph), it is possible that some of the checks above take some time. Still 8 minutes seems excessive, even for very large graphs. Can you please report your graph size? Like what is the number of nodes of your graph if you dump it with ``conan graph info --format=json > graph.json``?

If you can also share your full output, that might help to understand what is happening.
It would also help to report the timings when things are already installed in the cache, what is the time of a second ``conan install``?

Then, there are mechanisms to avoid this, like:
```
$ conan config list skip
...
tools.graph:skip_build: (Experimental) Do not expand build/tool_requires
```

This ``tools.graph:skip_build`` allows to completely prune the tool-requires. But I wouldn't consider this yet, it would be important to understand first what is happening.

**Comment by mpll-d** [comment]
First of all, thanks for the swift answer.
I'll try to address the different questions: 
- I have 89 nodes on that, which is not as huge as the thousands you mention.
- Out of the 8 minutes I mentioned, it seems to be about 50/50 time spent on `Computing dependency graph` vs the download of the dependencies
- A second run is almost instantaneous in regarding the `Computing dependency graph` step.
- During the `Computing dependency graph` step, most of the time is spent "thinking" (~1min) before displaying:
```
ninja/1.12.1: Not found in local cache, looking in remotes...
ninja/1.12.1: Checking remote: conancenter
ninja/1.12.1: Downloaded recipe revision fd583651bf0c6a901943495d49878803
```
Where ninja typically is a `tool_requires` package defined with a revision range.
So I guess the time is spent trying to resolve this version / revision for `ninja`.

I think I also got a fix, as I have been reading through the documentation [here](https://docs.conan.io/2/devops/using_conancenter.html). I saw that I should store the revs of the conancenter recipes I use on my own server, so I only store what I need, which should be much faster.
Also, adding a lockfiles makes it way faster. Though the issue with that is that I'm not sure how I should handle and maintain lockfiles. I'm building multiple packages and it seems a bit complicated to maintain consistent lockfiles around different repos, if I'm to commit them there. 
Is there a best practice on how to store. sync & use lockfiles for a poly-repo codebase?

Thanks for your help!

**Comment by memsharded** [comment]
> I have 89 nodes on that, which is not as huge as the thousands you mention.
Out of the 8 minutes I mentioned, it seems to be about 50/50 time spent on Computing dependency graph vs the download of the dependencies

That is definitely not large, there are users with thousands of nodes. 4 minutes just resolving the dependency graph is very weird.
A way to measure the performance without the installation part, would be, on a blank cache, do a ``conan graph info`` command with the same arguments. I can typically resolve and download all binaries for the "tensorflow-lite" example (around 50 deps, some of them large ones), a few seconds. For example:
```
 conan graph info --requires="opencv/[*]" --requires="tensorflow-lite/[*]" --format=json > graph.json
```
Takes around 30 seconds, with a blank cache, using only ConanCenter as remote. I guess that you are using ConanCenter only? If you have your own server, can you test using just your server (pre-uploading the packages there). I often do tests with a local ArtifactoryCE (from our downloads page) running on my machine to investigate issues and rule out client or server issues.

It is true that it seems that ConanCenter is slower than usual, we are investigating it, but still seems a different order of magnitude.

> During the Computing dependency graph step, most of the time is spent "thinking" (~1min) before displaying:

For a single dependency? just for the recipe-revision? That looks weird, maybe this is related to the issues in ConanCenter, and is affecting some packages more than others. I'll let the team know, and keep you updated.

Regarding the lockfiles questions, maybe lets keep this one focus on the performance issue until we understand it properly, because it shouldn't be that slow, and if you want you can open new tickets for the lockfiles and any other question you might have. Thanks for your feedback.

**Comment by memsharded** [comment]
Hi @mpll-d 

ConanCenter issues that could have affected performance were solved the 20th of Dec.
Is this still an issue? Do you have further feedback? As commented above, it is not expected that it is that slow to resolve.

**Comment by mpll-d** [comment]
Hello and sorry for the delayed answer,
Thanks again for your investigation and fix.

Unfortunately, from what I have seen, the issue of long resolve times was still present earlier this week, as I tried to install conan packages that require to resolve version ranges.
In the meantime, I have disabled the conancenter remote and done a `conan-center-index` fork and setup self-hosted packages. This way it is much faster and much more robust in terms of maintenance, so I'm fine with it.

Another thing that might have played a role in this sluggishness is that, parallel to the conancenter remote, I was having another in-house Conan repository enabled. That one was kind of bloated (long time without cleanup, so a lot of outdated packages stored there). Maybe this one could have slowed down a lot the resolution of the version ranges? I did not consider it nor mention it, because the actual packages that were slowing me down with version ranges did not exist at all on that remote. I don't know if your resolution algorithm is able to immediately recognize that, and skip searching further on that remote, of if this remote may have slowed down the whole process.

**Comment by memsharded** [comment]
> In the meantime, I have disabled the conancenter remote and done a conan-center-index fork and setup self-hosted packages. This way it is much faster and much more robust in terms of maintenance, so I'm fine with it.

Yes, this is definitely recommended for many production cases anyway, as pointed out in the docs you linked above, sounds good.


> Another thing that might have played a role in this sluggishness is that, parallel to the conancenter remote, I was having another in-house Conan repository enabled. That one was kind of bloated (long time without cleanup, so a lot of outdated packages stored there). Maybe this one could have slowed down a lot the resolution of the version ranges?

Yes, it is definitely possible that other remote was causing the slowness. Specially if the repository was very bloated and not optimized (it is possible to have huge Artifactory repositories and still resolve fast, we have seen them many times in big enterprise users, but it is true that it might also require some performance tuning around the server setup, the DB, etc). 
Conan will not discard any remote, as this can introduce confusing behaviors or even errors in the pipelines. So when resolving dependencies in the remotes, Conan will reach to every remote and get a response or fail completely if the remote doesn't work, but if it works but just slowly, Conan will get for that remote response before proceeding to the next one, as the remote ordering can also be relevant).

**Comment by memsharded** [comment]
Hi @mpll-d 

Any further feedback or question here? It seems likely that the bloated repo was being the cause, but please let us know if otherwise.
Thanks!

**Comment by mpll-d** [comment]
Hello @memsharded,
It looks like it was indeed related to the bloated repository, many thanks for the support: things are now running smoothly.

**Comment by memsharded** [comment]
Great, thanks for the feedback!
